compiler/porcelain/src/lib.rs

```diff
@@ -75,29 +75,16 @@ pub fn compile(
     source_path: String,
     opts: CompileOpts,
 ) -> Result<CodeObject, CompileError> {
-    macro_rules! try_parse {
-        ($x:expr) => {
-            match $x {
-                Ok(x) => x,
-                Err(e) => return Err(CompileError::from_parse(e, source, source_path)),
-            }
-        };
-    }
-    let res = match mode {
-        compile::Mode::Exec => {
-            let ast = try_parse!(parser::parse_program(source));
-            compile::compile_program(ast, source_path, opts)
-        }
-        compile::Mode::Eval => {
-            let statement = try_parse!(parser::parse_statement(source));
-            compile::compile_statement_eval(statement, source_path, opts)
-        }
-        compile::Mode::Single => {
-            let ast = try_parse!(parser::parse_program(source));
-            compile::compile_program_single(ast, source_path, opts)
-        }
+    let mode = match mode {
+        compile::Mode::Exec => parser::Mode::Module,
+        compile::Mode::Eval => parser::Mode::Expression,
+        compile::Mode::Single => parser::Mode::Interactive,
+    };
+    let ast = match parser::parse(source, mode) {
+        Ok(x) => x,
+        Err(e) => return Err(CompileError::from_parse(e, source, source_path)),
```

Comment:
```
let ast = parser::parse(source, mode).map_err(|e| CompileError::from_parse(e, source, source_path))?;
```

Reply:
`source_path` is moved into `from_parse()`, and rust doesn't connect the closure to the early return, so I have to manually do the map_err.